### ISSUE TITLE
404s and 500s

### ISSUE THREAD

**Issue by smashwilson** [deconst/deconst-docs#35]
We should have an answer for configurable per-domain 404 and 500 layouts.

**Comment by jrperritt** [comment]
If we use one directory per domain beneath `layouts`, I think each domain can have a single 404.hbs and 500.hbs that get returned on errors. Otherwise I think we'd have to know ahead of time the location of the desired 404/500 template for each domain. Maybe also have default templates at the same level as `layouts` in case a domain doesn't specify them (404.default.hbs/500.default.hbs).

**Comment by smashwilson** [comment]
That all sounds :+1: to me. I don't think we'd need the `.default.` bit in the default layout names though; those can just be templates called `404`, `500`, what have you in the top-level directory.

Something like this:

```
# Control repository
layouts/
  404.hbs # Default 404 error page
  500.hbs  # Default 500 error page
  developer.rackspace.com/
    404.hbs # Special 404 page for d.r.c. 
    500.hbs # Special 500 page for d.r.c.
```

Also, on reflection, they _should_ still be handlebars templates, because we'll still want to be able to re-use other layout snippets (`{{{ common/header.hbs }}}`, etc); we just won't have `{{{ metadata.body }}}` anywhere.

**Comment by jrperritt** [comment]
We can handle custom 404s just fine in the Flask request handler, but I'm not sure we can do custom (per-domain) 500s. We can create a custom 500 error handler, but it would be part of the application (not part of the request handler), and so it won't have any request information like the domain.

**Comment by smashwilson** [comment]
Well - you're talking about missing or bad _layouts_, which is part of this... but really what we want to do is show custom error pages to the end user from the _presenter_.

In that case, we should render a 500 "layout" if any of the presenter calls fails for a reason that isn't "resource missing," like:
- A 500 from the mapping service
- A 500 from the content service
- An error from MongoDB
- A 500 from the layout service for whatever reason
- Template expansion failed (maybe? we could probably soft-fail here and return the partial render if we can)

Similarly, we should render a 404 from the presenter if any of these happen:
- The mapping service returns a 404 (domain isn't found)
- The content service returns a 404 (no metadata envelope found)
- Layout service returns a 404 (requested layout not found)

Basically, we should be making some kind of decision with [this `err` value](https://github.com/deconst/presenter/blob/master/src/content.js#L117) to distinguish between these cases and responding appropriately, possibly by making another request to the layout service for the correct error template.

**Comment by jrperritt** [comment]
True, but we'd know if the content or mapping service failed immediately. That we can plan for. I'm talking about something going wrong in the layout-service that we can't anticipate.

**Comment by smashwilson** [comment]
Well, I'm assuming that Flask will already set the response's HTTP status to 500 if there's an uncaught exception in a handler, and presumably there's a way to recover that from the `err` variable in the presenter? That should be all we need from the layout handler in that case, right?

(I guess we'd probably still need to fall back to a hardcoded string or something if the request to retrieve the 500 page _itself_ fails.)

**Comment by jrperritt** [comment]
Done in deconst/layout-service#13 and deconst/presenter#18
